### PR TITLE
tests: Migrate more tests to variable chunk size parsing

### PR DESCRIPTION
**Pull request by hartwork**
Follow up to #755 in some sense

CC @Snild-Sony

### CODE REVIEW

**Review by Snild-Sony — APPROVED** [review]
LGTM. If the tests pass, go for it.

**Comment by hartwork** [comment]
@Snild-Sony a few are left (that would need more work) and a fix like bae30695a0887c17f8733cef78816d98d7f3e62a just turned out needed to plug undefined behavior. My goal was to go as far as cheaply possible for the moment.

File: expat/tests/common.c [hunk 1]
```diff
@@ -185,14 +185,16 @@ _XML_Parse_SINGLE_BYTES(XML_Parser parser, const char *s, int len,
   if (chunksize > 0) {
     // parse in chunks of `chunksize` bytes as long as possible
     for (; offset + chunksize < len; offset += chunksize) {
-      enum XML_Status res = XML_Parse(parser, s + offset, chunksize, XML_FALSE);
+      const char *const start = (s == NULL) ? NULL : (s + offset);
+      enum XML_Status res = XML_Parse(parser, start, chunksize, XML_FALSE);
```
Comment (Snild-Sony) [inline expat/tests/common.c:189]: It might be simpler to just move s forward after each parse: `s += chunksize`.

**Review by Snild-Sony — COMMENTED** [review]
Minor comment; the current version is also fine.

File: expat/tests/common.c (hunk 1 above)
Comment (hartwork) [reply expat/tests/common.c]: @Snild-Sony addressed in more detail at https://github.com/libexpat/libexpat/commit/7f0120dacbd08cacd6ce2e94247f2763eda78790#r131929667 , adjusted.

File: expat/tests/common.c [hunk 2]
```diff
@@ -181,18 +181,18 @@ enum XML_Status
 _XML_Parse_SINGLE_BYTES(XML_Parser parser, const char *s, int len,
                         int isFinal) {
   const int chunksize = g_chunkSize;
-  int offset = 0;
   if (chunksize > 0) {
     // parse in chunks of `chunksize` bytes as long as possible
-    for (; offset + chunksize < len; offset += chunksize) {
-      enum XML_Status res = XML_Parse(parser, s + offset, chunksize, XML_FALSE);
+    for (; chunksize <= len;
+         len -= chunksize, s = (s != NULL) ? (s + chunksize) : NULL) {
```
Comment (Snild-Sony) [inline expat/tests/common.c:187]: The NULL condition is unnecessary: chunksize is known to be non-zero here (we're only concerned about `NULL+0`, right?).

(also, len is known to be non-zero, and if someone sends in `s=NULL` and `len>0`, they _deserve_ the UB error.)

Comment (hartwork) [reply expat/tests/common.c:187]: > The NULL condition is unnecessary: chunksize is known to be non-zero here (we're only concerned about `NULL+0`, right?).

@Snild-Sony my patch was done with the deliberate "requirement" (or assumption) in mind to make `_XML_Parse_SINGLE_BYTES` support the full interface of `XML_Parse` but with chunking support added — a drop-in replacement — so that eventually we can replace _all_ calls to `XML_Parse` by `_XML_Parse_SINGLE_BYTES` except those two inside `_XML_Parse_SINGLE_BYTES`. The tests in `test_empty_parse` below are the ones that challenge that requirement and triggered undefined behavior:

https://github.com/libexpat/libexpat/blob/3a0c5d6ae53d64fd8f8abc0ff03dcd6cfb754030/expat/tests/basic_tests.c#L2730-L2758

And these include `NULL`. Of course we could exclude those cases but I would vote for making `_XML_Parse_SINGLE_BYTES` robust enough to handle all and then we don't need to check again for leftovers or document why particular calls are exempt. But maybe that's just my preference.

If you only concern left is to resolve `(s != NULL) ? x : y` to `x` then I would want to vote to leave that part in and go forward. One alternative would be to revert the whole patch to `_XML_Parse_SINGLE_BYTES` and make test `test_empty_parse` exercise plain `XML_Parse`. Both can be argued for.

What do you think?

Comment (Snild-Sony) [reply expat/tests/common.c:187]: > > The NULL condition is unnecessary: chunksize is known to be non-zero here (we're only concerned about `NULL+0`, right?).
> 
> @Snild-Sony my patch was done with the deliberate "requirement" (or assumption) in mind to make `_XML_Parse_SINGLE_BYTES` support the full interface of `XML_Parse` but with chunking support added — a drop-in replacement — so that eventually we can replace _all_ calls to `XML_Parse` by `_XML_Parse_SINGLE_BYTES`

I was under the impression that we were only trying to avoid adding `0`, since that was the specific error we had. My bad.

However, my other point about len=0 making sure we don't reach that code still stands:

> The tests in test_empty_parse below are the ones that challenge that requirement and triggered undefined behavior:  [...] And these include NULL.

But all of those also have len=0, which means the execution would not reach the line of code that we're discussing.

Comment (Snild-Sony) [reply expat/tests/common.c:187]: > If you only concern left is to resolve `(s != NULL) ? x : y` to `x`

It is. 

> then I would want to vote to leave that part in and go forward.

Making that choice is your privilege and your burden as maintainer. My words hold no power here. :)

Comment (hartwork) [reply expat/tests/common.c]: > > If you only concern left is to resolve `(s != NULL) ? x : y` to `x`
> 
> It is.

@Snild-Sony I think I may have found (and pushed) a compromise that gets rid of that and still doesn't do undefined behavior:
https://github.com/libexpat/libexpat/compare/421551302a37ddbdf7874c877070df2fa4d79567..6b5691bcf8b93668d8733aa3e99510be6e717a3b
Any better? What do you think?

> > then I would want to vote to leave that part in and go forward.
> 
> Making that choice is your privilege and your burden as maintainer. My words hold no power here. :)

Let's try with consensus first. Maybe we're already close to consensus now.

Comment (Snild-Sony) [reply expat/tests/common.c]: That part looks good now, thanks.

File: expat/tests/common.c [hunk 3]
```diff
@@ -191,18 +192,18 @@ enum XML_Status
 _XML_Parse_SINGLE_BYTES(XML_Parser parser, const char *s, int len,
                         int isFinal) {
   const int chunksize = g_chunkSize;
-  int offset = 0;
   if (chunksize > 0) {
     // parse in chunks of `chunksize` bytes as long as possible
-    for (; offset + chunksize < len; offset += chunksize) {
-      enum XML_Status res = XML_Parse(parser, s + offset, chunksize, XML_FALSE);
+    for (; chunksize <= len; len -= chunksize, s += chunksize) {
```
Comment (Snild-Sony) [inline expat/tests/common.c:197]: Sorry, I missed this previously:

Why `chunksize <= len`? It means the `XML_Parse()` at line 206 will have len=0 if the `len` arg is divisible by `chunksize`, essentially adding an extra `XML_Parse()` call that was not previously there.

Note that for chunksize 1, len will always be divisible, so we'll never make a `len=1 isFinal=1` call, just some amount of `len=1 isFinal=0` and the concluding `len=0 isFinal=1`.

It probably doesn't matter in practice for now, but could potentially help hide problems in the future.

File: expat/tests/common.c [hunk 4]
```diff
@@ -191,18 +192,18 @@ enum XML_Status
...
+      enum XML_Status res = XML_Parse(parser, s, chunksize, XML_FALSE);
       if (res != XML_STATUS_OK) {
         return res;
       }
+      assert(s != NULL); // or "s += chunksize" would be undefined behavior
```
Comment (Snild-Sony) [inline expat/tests/common.c:202]: I think an `assert(s != NULL || len == 0)` or `assert(! (s==NULL && len > 0))` at the start of the function or above the loop might be clearer, but this is also fine.

File: expat/tests/common.c (hunk 3 above)
Comment (hartwork) [reply expat/tests/common.c:197]: @Snild-Sony nice catches, maybe that's the last drop needed to reveal this is the wrong approach altogether. Let's flip this around and make `_XML_Parse_SINGLE_BYTES` reject pathological input like `s == NULL` loudly so that the tests in `test_empty_parse` have to resort to plain `XML_Parse` and keep testing the raw actual `XML_Parse` API. Pushing in a second…

File: expat/tests/common.c (hunk 4 above)
Comment (hartwork) [reply expat/tests/common.c:202]: @Snild-Sony I went even further with the 180° turn approach from https://github.com/libexpat/libexpat/pull/787#discussion_r1388499466 now. Pushing in a second…

File: expat/tests/common.c (hunk 3 above)
Comment (Snild-Sony) [reply expat/tests/common.c]: I just wanted it to be `chunksize < len`  instead, then it would've been perfect... :sob:

File: expat/tests/common.c [hunk 5]
```diff
@@ -190,6 +191,10 @@ _xml_failure(XML_Parser parser, const char *file, int line) {
 enum XML_Status
 _XML_Parse_SINGLE_BYTES(XML_Parser parser, const char *s, int len,
                         int isFinal) {
+  // This ensures that tests have to run pathological parse cases
+  // (e.g. when `s` is NULL) against plain XML_Parse rather than
+  // chunking _XML_Parse_SINGLE_BYTES.
+  assert((parser != NULL) && (s != NULL) && (len >= 0));
```
Comment (Snild-Sony) [inline expat/tests/common.c:197]: I preferred where we were going with the other approach, but this is also fine (as long as there are no tests that need to do `if NULL Parse else Parse_SINGLE`, but that doesn't seem to be the case right now).

File: expat/tests/common.c (hunk 3 above)
Comment (hartwork) [reply expat/tests/common.c]: @Snild-Sony I have re-added a refactoring of `_XML_Parse_SINGLE_BYTES` using `chunksize < len` as a dedicated commit just for you now :smiley: The assertion still blocks pathological input altogether.

Comment (hartwork) [reply expat/tests/common.c]: @Snild-Sony PS: I should note that GitHub web UI is behind in displayed pushed changes at the moment. Branch https://github.com/libexpat/libexpat/commits/tests-more-chunk-size-coverage has the commit but the pull request does not show it :scream: PPS: As with all status pages, https://www.githubstatus.com/ is all green, lovely.

**Review by Snild-Sony — APPROVED** [review]
